src/Telemetry/AzureDevOpsClient/AzureDevOpsClient.cs

```diff
@@ -177,49 +178,64 @@ private async Task<string> GetTimelineRaw(string project, int buildId, string id
+        Queue<string> lineCache = new();
+        // need to check if a new line will be loaded when the cache is full, it shouldn't because the first condition will tell it to exit the loop
+        while ((line = await reader.ReadLineAsync()) != null && lineCache.Count < regexes.Count())
         {
-            if (TryMatchRegex(line, regexes[regexIndex], out string? result))
-            {
-                // We need to cache lines to check the regex sequence again in case of a miss.
-                // We don't want to cache the line that was already matched by the first regex
-                if (regexIndex > 0)
-                {
-                    lineCache.Enqueue(line);
-                }
+            lineCache.Enqueue(line);
+        }
 
-                regexIndex++;
+        // Check if we didn't even have enough lines to fill the cache
+        if (lineCache.Count < regexes.Count())
+        {
+            return null;
+        }
 
-                if (regexIndex == regexes.Length)
-                {
-                    return result;
-                }
+        string? result;
+        do
+        {
+            result = CheckLineCache(lineCache, regexes);
+
+            if (result != null)
+            {
+                return result;
             }
-            else if (regexIndex > 0)
+
+            lineCache.Dequeue();
+            if (line != null)
             {
-                // Add the line that we missed on to the cache, it might fit the regex sequence when we go over the cache
                 lineCache.Enqueue(line);
-                regexIndex = 0;
+            }
+        }
+        while ((line = await reader.ReadLineAsync()) != null);
 
-                // We missed, time to go back to the cache and look if the sequence there will match
-                while (lineCache.Count > 0)
-                {
-                    var cachedLine = lineCache.Dequeue();
-                    if (TryMatchRegex(cachedLine, regexes[regexIndex], out _))
-                    {
-                        regexIndex++;
-                    }
-                    else
-                    {
-                        regexIndex = 0;
-                    }
-                }
+        // This will return the value if it finds something in the last cache, or null if not
+        return CheckLineCache(lineCache, regexes);
+    }
+
+    private string? CheckLineCache(IEnumerable<string> lineCache, IEnumerable<Regex> regexes)
+    {
+        string? result = null;
+        bool ok = true;
+
+        using var cacheEnum = lineCache.GetEnumerator();
+        using var regexEnum = regexes.GetEnumerator();
+        while (cacheEnum.MoveNext() && regexEnum.MoveNext())
+        {
+            ok &= TryMatchRegex(cacheEnum.Current, regexEnum.Current, out result);
+
+            if (!ok)
+            {
+                break;
             }
         }
 
+        if (ok)
+        {
+            return result;
+        }
+
         return null;
     }
```

Comment:
You could also use `Zip()` and `All()` maybe?

Reply:
Yup, I didn't think `All()` would work with the `out`, but it does